syntax = "proto3";
package pt.ulisboa.tecnico.tuplespaces.centralized.contract;

// The following option causes top-level messages, enums, and services to be defined at the package level, 
// rather than inside an outer class named after the .proto file.
option java_multiple_files = true;


message PutRequest {
  int32 clientId = 1;
  string newTuple = 2;
}
message PutResponse {
}

message ReadRequest {
  int32 clientId = 1;
  string searchPattern = 2;
}
message ReadResponse {
  string result = 1;
}

message TakeRequest {
  int32 clientId = 1;
  string searchPattern = 2;
}

message TakeResponse {
  string result = 1;
}

message getTupleSpacesStateRequest {
  int32 clientId = 1;
}
message getTupleSpacesStateResponse {
  repeated string tuple = 1;
}

service TupleSpaces {
  rpc put(PutRequest) returns (PutResponse);
  rpc read(ReadRequest) returns (ReadResponse);
  rpc take(TakeRequest) returns (TakeResponse);
  rpc getTupleSpacesState(getTupleSpacesStateRequest) returns (getTupleSpacesStateResponse);
}
